### PR TITLE
feat(auth): redirected login

### LINKED ISSUE #511: Support 2FA and alternative login flows
Currently our adapter relies on username and password.  This is sufficient for basic SAS installs however there are many sites using 2FA or alternative authentication providers that can result in multiple hops away from the SASLogon page (and don't involve submitting a username/password to SAS).

We therefore need a new login flow that can handle such scenarios.  The flow would go like this:

* the sasjsconfig will contain a new attribute (perhaps: `loginMechanism`) with values such as: `USERPASS` (default) or `REDIRECT` (suggested).
* If `loginMechanism: 'REDIRECT'` then we need to open a separate SASLogon window (or iframe) so the user can perform whichever login flow has been configured within SAS.  This may or may not redirect back to the application.
* In the background, the adapter needs to poll SAS every 1000ms with `isLoggedIn()` to check login status.  
* This process should work for the submitted request promise - so if a request is submitted, the user is not logged in, we perform the redirect flow and the original request is submitted once authenticated.
* This flow should work for both SAS 9 and Viya

More information on authentication: https://communities.sas.com/t5/Ask-the-Expert/How-Do-I-Authenticate-to-SAS-Viya/ta-p/761328

This process can be tested in two ways (after a logout):

* By entering login details into the new window OR
* By entering login details in a _different_ window, eg /SASStudio or /SASStudioV

### PR DESCRIPTION
**Pull request by saadjutt01**
## Issue

Closes #511 

## Intent

Implement login through new popup window.

## Implementation

Added `LoginMechanism` to `SASjsConfig.ts`

`SASjs.ts` changes:
- Defined returnType for `SASjs.login()`
- Added optional `LoginOptions` 

`AuthManager.ts` changes:
- Added another login function `redirectedLogin`
- Extracted code from `checkSession` to new private function `fetchUserName` to make it DRY

## Checks

No PR (that involves a non-trivial code change) should be merged, unless all items below are confirmed!  If an urgent fix is needed - use a tar file.

- [x] All `sasjs-cli` unit tests are passing (`npm test`).

Mac - Mocked tests:
image

Mac - Server tests:
image

Windows - Mocked tests:
image

Windows - Server tests:
image

- [x] All `sasjs-tests` are passing (instructions available [here](https://github.com/sasjs/adapter/blob/master/sasjs-tests/README.md)).

`SASVIYA`
image

`SAS9` 
image

- [x] [Data Controller](https://datacontroller.io) builds and is functional on both SAS 9 and Viya
Tested with `SASVIYA`
Tested with `SASVIYA` having login mechanism 'Redirected'
Tested with `SAS9`

### CODE REVIEW

File: src/types/Login.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export interface LoginOptions {
+  onLoggedOut?: () => Promise<boolean>
+}
+
+export interface LoginReturn {
```
Comment (krishna-acondy) [inline src/types/Login.ts:5]: I'd probably call this `LoginResult`.

File: src/utils/loginPrompt/index.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  Object.keys(domIDs).forEach((id) => {
+    const elem = document.getElementById(id)
+    elem?.parentNode?.removeChild(elem)
+  })
+
+  document.body.style.overflow = 'auto'
+}
+
+const cssContent = `
+.popUp {
```
Comment (krishna-acondy) [inline src/utils/loginPrompt/index.ts:66]: Have we checked this at a few different resolutions? I see you've mostly used percentages for widths and heights so we should be okay, but I just want to be completely sure.

File: src/auth/verifyingPopUpLoginSASVIYA.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { delay } from '../utils'
+
+export async function verifyingPopUpLoginSASVIYA(loginPopup: Window) {
```
Comment (krishna-acondy) [inline src/auth/verifyingPopUpLoginSASVIYA.ts:3]: I'd call this file and function `verifySasViyaLoginPopup`.

File: src/auth/verifyingPopUpLoginSAS9.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { delay } from '../utils'
+
+export async function verifyingPopUpLoginSAS9(loginPopup: Window) {
```
Comment (krishna-acondy) [inline src/auth/verifyingPopUpLoginSAS9.ts:3]: I'd probably call this file and function `verifySas9Login`.

File: src/auth/openWebPage.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+interface windowFeatures {
+  width: number
+  height: number
+}
+
+export async function openWebPage(
+  url: string,
+  windowName: string = '',
+  { width, height }: windowFeatures,
```
Comment (krishna-acondy) [inline src/auth/openWebPage.ts:11]: I think it'll be good to provide defaults for `width` and `height` here.

File: src/auth/AuthManager.ts [hunk 6]
```diff
@@ -167,11 +197,34 @@ export class AuthManager {
...
   }
 
+  private async fetchUserName(): Promise<{
+    isLoggedIn: boolean
+    userName: string
+  }> {
+    //For VIYA we will send request on API endpoint. Which is faster then pinging SASJobExecution.
+    //For SAS9 we will send request on SASStoredProcess
+    const url =
+      this.serverType === 'SASVIYA'
```
Comment (krishna-acondy) [inline src/auth/AuthManager.ts:212]: We should check against the `ServerType` enum from `@sasjs/utils/types`.

**Comment by krishna-acondy** [comment]
Looks mostly good apart from some minor changes I've suggested.

It'd be good to also see some unit tests for the new stuff added to `AuthManager`.

File: src/types/Login.ts (hunk 1 above)
Comment (saadjutt01) [reply src/types/Login.ts:5]: fixed.

File: src/auth/verifyingPopUpLoginSASVIYA.ts (hunk 3 above)
Comment (saadjutt01) [reply src/auth/verifyingPopUpLoginSASVIYA.ts:3]: fixed.

File: src/auth/openWebPage.ts (hunk 5 above)
Comment (saadjutt01) [reply src/auth/openWebPage.ts:11]: fixed.

File: src/auth/AuthManager.ts (hunk 6 above)
Comment (saadjutt01) [reply src/auth/AuthManager.ts:212]: fixed.

File: src/auth/openWebPage.ts (hunk 5 above)
Comment (saadjutt01) [reply src/auth/openWebPage.ts:11]: fixed.

File: src/auth/verifyingPopUpLoginSASVIYA.ts (hunk 3 above)
Comment (saadjutt01) [reply src/auth/verifyingPopUpLoginSASVIYA.ts:3]: fixed.

File: src/auth/verifyingPopUpLoginSAS9.ts (hunk 4 above)
Comment (saadjutt01) [reply src/auth/verifyingPopUpLoginSAS9.ts:3]: fixed.

**Review by krishna-acondy — APPROVED** [review]

**Review by allanbowe — APPROVED** [review]

**Pull request merged**